common/mqtt_client/src/lib.rs

```diff
@@ -490,6 +490,13 @@ impl Message {
     pub fn qos(self, qos: QoS) -> Self {
         Self { qos, ..self }
     }
+
+    fn trim_if_null_terminated(payload: &mut Vec<u8>) -> &mut Vec<u8> {
+        if !payload.is_empty() && payload[payload.len() - 1].eq(&0) {
```

Comment:
1. Don't make this a function belonging to `Message`. It has nothing to do with a message. it just drops the NUL of a byte vector.

2. The implementation does one copy too much. Each time, it creates a clone of the vector in `to_vec`, which is unneccesarry.

Use this function instead:

```rust
fn trim_null_terminated(mut vec: Vec<u8>) -> Vec<u8> {
    if let Some(0) = vec.as_slice().last() {
        vec.pop();
    }
    vec
}
```

Reply:
And then:

```rust
   payload: trim_null_terminated(payload.into())
```